[misc] RequestParser: restrict imageName to an allow list and add tests

das7pad opened the pull request:
### Description

For https://github.com/overleaf/issues/issues/3171

We may want to downgrade the server error (500) to a client error (400) when a compile request fails the validation. This should be done in a different PR as it likely involved changes in web.

#### Related Issues / PRs

https://github.com/overleaf/issues/issues/3171

https://github.com/overleaf/google-ops/pull/1113 to inject the environment variable (optional, this PR can get deploy independently).

#### Potential Impact

High. Users with bad entries in the db might have to reset the imageName from the ui.

#### Manual Testing Performed

- added unit and acceptance tests

#### Deployment checklist

- [ ] https://github.com/overleaf/google-ops/pull/1113

briangough commented on app/js/RequestParser.js:
```diff
@@ -61,7 +61,7 @@ module.exports = RequestParser = {
         'imageName',
         compile.options.imageName,
-        { type: 'string' }
+        { type: 'string', validValues: settings.allowedImageNamesFlat }
```
Add a default of settings.clsi.docker.image to handle empty values?

das7pad replied:
I am not sure what you mean here.

_Empty values simply skip the validation logic and I'd rather keep it that way:_

https://github.com/overleaf/clsi/blob/b33734bab64beee8f06327093f9b4bf67c2d5eec/app/js/RequestParser.js#L190-L196

briangough replied:
You're right, that's not needed.

das7pad commented:
Thanks for the feedback!

> * [ ]  Use `allowedImageNames` instead of `allowedImageNamesFlat`, it's more idiomatic.

There is already a `allowedImageNames` field in the code base of overleaf: web has a map at this setting. To differentiate between the two, I chose to use a slightly different name which denotes that is a _flat_ version.

> * [ ]  The setting only applies to docker so it can be inside the process.env.DOCKER_RUNNER conditional in settings.default.js as a `settings.clsi.docker.allowedImageNames property`.

Good point, I'll move it.

> * [ ]  Check the image name at the point where it is used in DockerRunner, to avoid having multiple checks in the controller.   I think it is ok to check in the RequestParser as well.

I'm not sure about this one -- any validation should happen at the controller level and the manager/handler/runner should just eat what was passed. E.g. we do not validate the mounts at the runner level, but they could contain `../` and lead to unexpected mounts. https://github.com/overleaf/clsi/blob/6b69e26de31ff4b739b47f0fe12cf3da81930861/app/js/DockerRunner.js#L78

briangough commented:
> > * [ ]   Use `allowedImageNames` instead of `allowedImageNamesFlat`, it's more idiomatic.
> 
> There is already a `allowedImageNames` field in the code base of overleaf: web has a map at this setting. To differentiate between the two, I chose to use a slightly different name which denotes that is a _flat_ version.

I don't think it's a problem in practice for the setting to have a different type here and in web, but if you are looking for an alternative name `allowedImages` would be fine.

> > * [ ]   Check the image name at the point where it is used in DockerRunner, to avoid having multiple checks in the controller.   I think it is ok to check in the RequestParser as well.
> 
> I'm not sure about this one -- any validation should happen at the controller level and the manager/handler/runner should just eat what was passed. E.g. we do not validate the mounts at the runner level, but they could contain `../` and lead to unexpected mounts.

My main concern is that these kind of checks should not be specific to individual endpoints, otherwise it is easily forgotten if a new endpoint is added or modified to accept an image (as may happen for synctex https://github.com/overleaf/issues/issues/3170). 

DockerRunner would be a reasonable place to put that check - at the point of use (and it would be a good idea to have some checks on the volume paths there too).   

Checking in the controller as well is good, but I think the check should be made in a more general way.

das7pad commented:
I've addressed the review feedback in https://github.com/overleaf/clsi/pull/183/commits/ee0e8066d3c6d6e3f4041e2364c631fe451558be

---

> and it would be a good idea to have some checks on the volume paths there too

Let's do this in a followup-PR.

briangough commented:
See https://github.com/overleaf/google-ops/pull/1113#pullrequestreview-440662996 about a problem with overriding the image name.

das7pad commented on app/js/DockerRunner.js:
```diff
@@ -91,6 +91,13 @@ module.exports = DockerRunner = {
       image = `${Settings.texliveImageNameOveride}/${img[2]}`
     }
 
+    if (
+      Settings.clsi.docker.allowedImages &&
+      !Settings.clsi.docker.allowedImages.includes(image)
+    ) {
+      return callback(new Error('image not allowed'))
+    }
```
We could move this check above the override and provide the quay names in the env var.

briangough replied:
Yes, that would be fine.

das7pad replied:
Done in https://github.com/overleaf/clsi/pull/183/commits/0cecf2656981141e8951ac8aafca6a1c07c972a9. google-ops change: https://github.com/overleaf/google-ops/compare/d763fc0a1cd1832500f00fe84200ba55e51cb40b..87e5057b926400929f2efc97509529d1e221a444
